packages/SwingSet/src/kernel/state/vatKeeper.js

```diff
@@ -115,19 +123,45 @@ export function makeVatKeeper(
   }
 
   function setReachableFlag(kernelSlot) {
-    // const { type } = parseKernelSlot(kernelSlot);
+    const { type } = parseKernelSlot(kernelSlot);
     const kernelKey = `${vatID}.c.${kernelSlot}`;
-    const { vatSlot } = parseReachableAndVatSlot(kvStore.get(kernelKey));
-    // const { allocatedByVat } = parseVatSlot(vatSlot);
+    const { isReachable, vatSlot } = parseReachableAndVatSlot(
+      kvStore.get(kernelKey),
+    );
+    const { allocatedByVat } = parseVatSlot(vatSlot);
     kvStore.set(kernelKey, buildReachableAndVatSlot(true, vatSlot));
+    // increment 'reachable' part of refcount, but only for object imports
+    if (!isReachable && type === 'object' && !allocatedByVat) {
+      // eslint-disable-next-line prefer-const
+      let { reachable, recognizable } = getObjectRefCount(kernelSlot);
+      reachable += 1;
+      setObjectRefCount(kernelSlot, { reachable, recognizable });
+    }
   }
 
   function clearReachableFlag(kernelSlot) {
-    // const { type } = parseKernelSlot(kernelSlot);
+    const { type } = parseKernelSlot(kernelSlot);
     const kernelKey = `${vatID}.c.${kernelSlot}`;
-    const { vatSlot } = parseReachableAndVatSlot(kvStore.get(kernelKey));
-    // const { allocatedByVat } = parseVatSlot(vatSlot);
+    const { isReachable, vatSlot } = parseReachableAndVatSlot(
+      kvStore.get(kernelKey),
+    );
+    const { allocatedByVat } = parseVatSlot(vatSlot);
     kvStore.set(kernelKey, buildReachableAndVatSlot(false, vatSlot));
+    // decrement 'reachable' part of refcount, but only for object imports
+    if (
+      isReachable &&
+      type === 'object' &&
+      !allocatedByVat &&
+      kernelObjectExists(kernelSlot)
```

Comment:
Pondering this `kernelObjectExists` check: it's present here but not in `setReachableFlag`.  On the one hand, I suppose you couldn't be setting the flag on something that doesn't exist.  On the other hand, why would you be trying to clear it if it doesn't exist?  My intuition is that the check might actually still be right (e.g., don't decrement below 0), but I think the asymmetry needs some explanation.

Reply:
Good observation. Yes, your first point is correct: we call `setReachableFlag` only from within `mapVatSlotToKernelSlot`/`mapKernelSlotToVatSlot`, just after we've added a thing to the c-list or seen an existing thing, so the kernel object exists. (Except when the `setReachable` flag is false, which we do when translating e.g. `dispatch.retireExport` into a vat, which means we've already deleted the kernel object and we need a vref to tell the vat about it, but don't want to resurrect the object).

We deliberately clear the reachable flag in three cases:
* the vat does `syscall.dropImports` (i.e. a Presence was dropped)
* the kernel does `dispatch.dropExports` (i.e. "hey vat please drop your Remotable")
* we delete a c-list entry, because of one of five cases:
  * the vat does `syscall.retireExports` on a previously-dropped vref (a Remotable was dropped)
  * the vat does `syscall.retireImports` on a previously-dropped (i.e. unreachable) vref (i.e. a Presence was dropped even harder)
  * the kernel does `dispatch.retireExports` on a previously-dropped vref ("hey vat, nobody can recognize your old Remotable")
  * the kernel does `dispatch.retireImports` on a previously-dropped vref ("hey vat, that Presence you think you would still recognize, it's gone, forget about it")
  * a vat is terminated, and we `deleteCListEntry` everything it used to have (reachable or not)

In the `drop` cases, we expect `isReachable` to be set upon entry, and we're clearing it. In the `retire` cases, we'd expect it to be clear upon entry (there *might* be a race which could allow something to re-set it first, I'm not sure right now). In the terminate-vat case, `isReachable` could be in either state. It seemed safest/simplest to use a `clearReachable()` that could handle either state, and only decrement the `reachable` counter if it was set upon entry, rather than have each of these cases call a different function.

In the `dispatch.retire` cases, the kernel object will have been deleted a moment earlier, so we need to tolerate the lack of a reachahle/recognizable count. Which means `deleteCListEntry` needs to tolerate a a missing kernel object. `deleteCListEntry` is responsible for the decref in the vat-termination case, so it must call `clearReachable()`, which means `clearReachableFlag()` must tolerate a missing kernel object (or `deleteCListEntry` must sense the missing kernel object and then refrain from calling `clearReachableFlag()`, but that seemed worse).

I'm working on adding counters to track all the ups and downs of reachable/recognizable states, and I'm running into the same sort of questions. The increment wants to live in one place, but the decrement happens somewhere seemlngly-unrelated. I'll take another look at the code and see if there's a better way to arrange it that would make it more symmetric.